### PR TITLE
SUBMARINE-1174. K8sSubmitter refactoring involving Notebook create and delete method

### PR DESCRIPTION
**Pull request by cdmikechen**
### What is this PR for?
The current K8sSubmitter contains a lot of CRD processing logic in addition to the implementation of the interface. 
The purpose of this PR is to split the CRD processing logic from the class as far as possible and merge it into the resource object for processing.

The current PR refactored part of the code so that we can see the changes of the code after refactoring.

### What type of PR is it?
Refactoring

### Todos
* [x] - Refactoring `K8sSubmitter`  involving Notebook create and delete method (Configmap/PVC/IstioVirtualService/NoteBookCR)
* [x] - Add k8s mock client/server for test  

### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-1174

### How should this be tested?
The refactored codes can use the original test case.

### Screenshots (if appropriate)
No

### Questions:
* Do the license files need updating? No
* Are there breaking changes for older versions? No
* Does this need new documentation? No

### CODE REVIEW

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/Configmap.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    init(namespace, name, datas);
+  }
+
+  private void init(String namespace, String name, Map<String, String> datas) {
+    /*
+      Required value
+      1. metadata.name
+      2. metadata.namespace
+      3. spec.data
+      4. spec.resources
```
Comment (lowc1012) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/Configmap.java:65]: Should be "3. data".
And I can't find the "spec.resources" in ConfigMap resource.

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/Configmap.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      LOG.error("K8s submitter: parse configmap object failed by " + e.getMessage(), e);
+      throw new SubmarineRuntimeException(e.getCode(), "K8s submitter: parse configmap object failed by " +
+              e.getMessage());
+    }
+  }
+
+  @Override
+  public Object create(K8sClient api) {
+    try {
+      return api.getCoreApi().createNamespacedConfigMap(this.getMetadata().getNamespace(), this, "true",
```
Comment (lowc1012) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/Configmap.java:100]: may produce NPE problem

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/Configmap.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      LOG.error("K8s submitter: replace configmap object failed by " + e.getMessage(), e);
+      throw new SubmarineRuntimeException(e.getCode(), "K8s submitter: replace configmap object failed by " +
+              e.getMessage());
+    }
+  }
+
+  @Override
+  public Object delete(K8sClient api) {
+    try {
+      return api.getCoreApi().deleteNamespacedConfigMap(this.getMetadata().getName(),
```
Comment (lowc1012) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/Configmap.java:129]: may produce NPE problem

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/Configmap.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+      throw new SubmarineRuntimeException(e.getCode(), "K8s submitter: create configmap failed by " +
+              e.getMessage());
+    }
+  }
+
+  @Override
+  public Object replace(K8sClient api) {
+    try {
+      // reset metadata to get resource version so that we can replace configmap
+      if (StringUtils.isBlank(this.getMetadata().getResourceVersion())) {
```
Comment (lowc1012) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/Configmap.java:113]: may produce NPE problem

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/PersistentVolumeClaim.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  @Override
+  public PersistentVolumeClaim read(K8sClient api) {
+    return this;
+  }
+
+  @Override
+  public V1PersistentVolumeClaim create(K8sClient api) {
+    try {
+      return api.getCoreApi().createNamespacedPersistentVolumeClaim(
+              this.getMetadata().getNamespace(), this, "true", null, null
```
Comment (lowc1012) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/PersistentVolumeClaim.java:76]: May produce NPE problem.

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/PersistentVolumeClaim.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  @Override
+  public Object replace(K8sClient api) {
+    return null;
+  }
+
+  @Override
+  public Object delete(K8sClient api) {
+    try {
+      return api.getCoreApi().deleteNamespacedPersistentVolumeClaim(
+              this.getMetadata().getName(), this.getMetadata().getNamespace(), "true",
```
Comment (lowc1012) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/model/common/PersistentVolumeClaim.java:94]: May produce NPE problem.

**Review by pingsutw — APPROVED** [review]

**Pull request closed without merging**